keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/KeelRepository.kt

```diff
@@ -0,0 +1,228 @@
+package com.netflix.spinnaker.keel.persistence
+
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.ResourceSpec
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactStatus
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactType
+import com.netflix.spinnaker.keel.api.artifacts.DebianArtifact
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.artifacts.DockerArtifact
+import com.netflix.spinnaker.keel.api.id
+import com.netflix.spinnaker.keel.constraints.ConstraintState
+import com.netflix.spinnaker.keel.core.api.EnvironmentArtifactPin
+import com.netflix.spinnaker.keel.core.api.EnvironmentArtifactVetoes
+import com.netflix.spinnaker.keel.core.api.EnvironmentArtifactsSummary
+import com.netflix.spinnaker.keel.core.api.PinnedEnvironment
+import com.netflix.spinnaker.keel.core.api.SubmittedDeliveryConfig
+import com.netflix.spinnaker.keel.core.api.UID
+import com.netflix.spinnaker.keel.diff.DefaultResourceDiff
+import com.netflix.spinnaker.keel.events.ResourceCreated
+import com.netflix.spinnaker.keel.events.ResourceEvent
+import com.netflix.spinnaker.keel.events.ResourceUpdated
+import com.netflix.spinnaker.keel.exceptions.DuplicateResourceIdException
+import java.time.Clock
+import java.time.Duration
+import org.slf4j.Logger
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.transaction.annotation.Propagation
+import org.springframework.transaction.annotation.Transactional
+
+/**
+ * A repository for interacting with delivery configs, artifacts, and resources.
+ */
+interface KeelRepository {
+  val clock: Clock
+  val publisher: ApplicationEventPublisher
+  val log: Logger
+
+  @Transactional(propagation = Propagation.REQUIRED)
+  fun upsertDeliveryConfig(submittedDeliveryConfig: SubmittedDeliveryConfig): DeliveryConfig
+
+  /**
+   * Validates that resources have unique ids, throws an exception if invalid
+   */
+  fun validate(config: DeliveryConfig) {
```

Comment:
This seems out of place on the interface, it mixes up model concerns with storage. It looks like it doesn't depend on other repository methods unless I'm missing something, so maybe it could be a top-level extension function in DeliveryConfig.kt or somewhere adjacent.

Otherwise LGTM, great work!

Reply:
Ah good point. I'll fix!